### PR TITLE
Unarchive Spark test jar for spark.read(ability)

### LINKED ISSUE #10875: [FEA] Access test files in resources from Spark source code when running Spark UT
**Is your feature request related to a problem? Please describe.**
Some Spark UT will try to read files in resources from Spark's code. So when we introduce Spark UT to plugin, we can't read those files directly.

For example, "SPARK-31716: inferring should handle malformed input" in `RapidsJsonSuite`  got following error (if included):

```
- SPARK-31716: inferring should handle malformed input *** FAILED ***
  org.apache.spark.sql.AnalysisException: Path does not exist: file:/home/haoyangl/spark-rapids/tests/src/test/resources/test-data/malformed_utf8.json
  at org.apache.spark.sql.errors.QueryCompilationErrors$.dataPathNotExistError(QueryCompilationErrors.scala:1011)
  at org.apache.spark.sql.execution.datasources.DataSource$.$anonfun$checkAndGlobPathIfNecessary$4(DataSource.scala:785)
  at org.apache.spark.sql.execution.datasources.DataSource$.$anonfun$checkAndGlobPathIfNecessary$4$adapted(DataSource.scala:782)
  at org.apache.spark.util.ThreadUtils$.$anonfun$parmap$2(ThreadUtils.scala:372)
  at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
  at scala.util.Success.$anonfun$map$1(Try.scala:255)
  at scala.util.Success.map(Try.scala:213)
  at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
  at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
  at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
  ...
```

I hope we can find a way to read test files in Spark's resources so we can really test related Spark UT cases.

**Describe the solution you'd like**

Gluten overrides testFile
```scala
  /** Returns full path to the given file in the resource folder */
  override protected def testFile(fileName: String): String = {
    getWorkspaceFilePath("sql", "core", "src", "test", "resources").toString + "/" + fileName
  }
```
where `getWorkspaceFilePath` in Spark
```scala
/**
 * Get a Path relative to the root project. It is assumed that a spark home is set.
 */
protected final def getWorkspaceFilePath(first: String, more: String*): Path = {
  if (!(sys.props.contains("spark.test.home") || sys.env.contains("SPARK_HOME"))) {
    fail("spark.test.home or SPARK_HOME is not set.")
  }
  val sparkHome = sys.props.getOrElse("spark.test.home", sys.env("SPARK_HOME"))
  java.nio.file.Paths.get(sparkHome, first +: more: _*)
}

```

Gluten is leveraging the system property "spark.test.home" (check [here](https://github.com/apache/incubator-gluten/blob/9c15d2e1e2f5cac4cf65c169c099f81c1ba1bc2e/.github/workflows/velox_docker.yml#L472)). For running Spark UT on shim 3.x.y, Gluten will prepare the docker container with a source folder filled with 3.x.y's code.

To do the same thing, we need to set up a Spark source folder before running Spark UT CI, and noted how to set up the env when running spark UT locally.

**Describe alternatives you've considered**
Those resource files are also included in jars, it is possible to read them from the jars if we know where the files are located.

**Additional context**
copy files to plugin pr: https://github.com/NVIDIA/spark-rapids/pull/10864
Spark UT RapidsJsonSuite issue: https://github.com/NVIDIA/spark-rapids/issues/10773

### PR DESCRIPTION
**Pull request by gerashegalov**
Closes #10875
Contributes to #10773
    
Unjar, cache, and share the test jar content among all test suites from the same jar

Test:
```bash
mvn package -Dbuildver=330 -pl tests -am -Dsuffixes='.*\.RapidsJsonSuite'
```

Signed-off-by: Gera Shegalov <gera@apache.org>

### CODE REVIEW

**Comment by gerashegalov** [comment]
build

File: tests/src/test/spark330/scala/org/apache/spark/sql/rapids/utils/RapidsSQLTestsBaseTrait.scala [hunk 1]
```diff
@@ -107,7 +116,36 @@ trait RapidsSQLTestsBaseTrait extends SharedSparkSession with RapidsTestsBaseTra
...
+  private def expandJar(jarPath: String): java.nio.file.Path = {
+    val jarFile = new java.io.File(jarPath)
+    val destDir = java.nio.file.Files.createTempDirectory(jarFile.getName + ".expanded")
+    logWarning(s"Registering $destDir for deletion on exit")
+    FileUtil.unZip(jarFile, destDir.toFile)
+    destDir
+  }
+
+  def sparkTestResourcesDir(testClass: Class[_]): java.nio.file.Path = {
+    val sparkTestClassResource = "/" + testClass.getSuperclass.getName.replace(".", "/") + ".class"
```
Comment (jlowe) [inline tests/src/test/spark330/scala/org/apache/spark/sql/rapids/utils/RapidsSQLTestsBaseTrait.scala:138]: Is just getting the superclass a safe assumption? We have some RAPIDS suite classes that derive from other RAPIDS classes before they derive from a Spark suite class.   Should this be looking up the class hierarchy for a  class without ".rapids" in the fully qualified classname?

Comment (gerashegalov) [reply tests/src/test/spark330/scala/org/apache/spark/sql/rapids/utils/RapidsSQLTestsBaseTrait.scala:138]: Good point, this works only for classes immediately deriving from a Suite in Spark repo

File: tests/src/test/spark330/scala/org/apache/spark/sql/rapids/utils/RapidsTestSettings.scala [hunk 2]
```diff
@@ -73,7 +73,6 @@ class RapidsTestSettings extends BackendTestSettings {
     .exclude("SPARK-23723: checking that the encoding option is case agnostic", KNOWN_ISSUE("https://github.com/NVIDIA/spark-rapids/issues/10773"))
     .exclude("SPARK-23723: specified encoding is not matched to actual encoding", KNOWN_ISSUE("https://github.com/NVIDIA/spark-rapids/issues/10773"))
     .exclude("SPARK-23724: lineSep should be set if encoding if different from UTF-8", KNOWN_ISSUE("https://github.com/NVIDIA/spark-rapids/issues/10773"))
-    .exclude("SPARK-31716: inferring should handle malformed input", KNOWN_ISSUE("https://github.com/NVIDIA/spark-rapids/issues/10773"))
     .exclude("SPARK-24190: restrictions for JSONOptions in read", KNOWN_ISSUE("https://github.com/NVIDIA/spark-rapids/issues/10773"))
```
Comment (thirtiseven) [inline tests/src/test/spark330/scala/org/apache/spark/sql/rapids/utils/RapidsTestSettings.scala:76]: Cases from `SPARK-23723` to this one `SPARK-24190` should be pass after this fix, can we remove them from exclude?

Comment (gerashegalov) [reply tests/src/test/spark330/scala/org/apache/spark/sql/rapids/utils/RapidsTestSettings.scala]: It worked, thank you!

**Comment by gerashegalov** [comment]
build

**Review by jlowe — APPROVED** [review]

**Pull request merged**